### ISSUE TITLE
Node labels

### ISSUE THREAD

**Issue by jroper** [akka/akka#29681]
This feature request might not actually require any code, depending on how it's done. Currently, Akka cluster nodes can have roles. There are currently no semantics in the names of roles, adding a particular role means nothing to Akka unless the end user has explicitly configured something to treat that role specially (in sharding, singleton, etc). In some discussions we've found that we might want something with a little more flexibility, something more akin to Kubernetes labels, where nodes can have key/value metadata attached to them, allowing reusable components to define semantics in the label names and read them.

A concrete example discussed was for the Knative weighted shard allocation strategy discussed [here](https://github.com/cloudstateio/cloudstate/issues/386) in Cloudstate. The cluster needs to know which revision a node is part of. The revision name could be passed as a role, however Cloudstate may use lots of different roles for different purposes, how would the Knative shard allocation strategy know which configured role corresponds to a nodes Knative revision name? If there was key value metadata associated with a node, then that might be done by looking up the value for `knative-revision` from that metadata.

Of course, this can be achieved today with no changes to Akka. We could in Cloudstate define our own convention of prefixing the knative revision name with `knative-revision=<name>` and passing that as a role, and also using a similar approach for other metadata. I've raised this issue firstly to discuss whether we should define a best practice or convention around this so that if reusable components were to start using this approach, they would use the same approach with the same labelling strategy, which would give users a consistent experience.

Additionally, Akka could provide some additional features to assist with this convention, the two main ones being:

* Parsing roles according to this convention and making them available as a multi valued map.
* Providing additional configuration mechanisms for adding roles according to this convention, this would greatly facilitate passing this metadata as system properties as currently the roles are defined in a HOCON array which are not easy to work with in system properties. For example, a nodes roles might be configured to be the superset of `akka.cluster.roles` as well as `akka.cluster.labels`, where the latter is a HOCON map with either string or string list values, which get formatted according to the convention to be roles.

**Comment by octonato** [comment]
@jroper, we have recently added an `app-version` settings. See https://github.com/akka/akka/pull/29546

Could that be used as the revision?

**Comment by jroper** [comment]
Not sure about that. The thing that concerns me is the fact that Akka is putting semantics on the app-version with regards to the orderability. While Knative does produce sequential revision numbers, they do not necessarily correlate to sequential version numbers. For example, Knative supports things like canary deploys and A/B deploys... in canary deploys, the highest revision number should not necessarily be considered the current version. In A/B deploys, you may have two different current revisions which should be equally considered the current version, one should not be favoured over the other for things like rebalancing and the like.

**Comment by patriknw** [comment]
> The thing that concerns me is the fact that Akka is putting semantics on the app-version with regards to the orderability.

That is used by the default allocation strategy to support smooth rolling updates by avoiding allocations to old nodes. As I understood it you had plans to write some other allocation strategy for Knative? Other allocation strategies may interpret the app-version in different ways to support other deploys. The original configured app-version String is preserved. If that works out we could incorporate it in Akka.

By the way, I hope you have seen that we have included a new allocation strategy with much faster rebalance: https://github.com/akka/akka/pull/29555 (to be released in upcoming 2.6.10 this week)

**Comment by ignasi35** [comment]
@jroper are you asking for a generic feature (key-value metadata ar node level) and you used the deployment scenarios (canary, A/B test) as possible use cases; or are you trying to solve the deployment scenarios issue and suggested `Node labels`?

In Kubernetes, labels and annotations are similar enough but meant for different things. Both are arbitrary Key/Value's but annotations are meant for application-internal usage while labels are exposed and used by kubernetes itself. At the same time, (I think) annotations are immutable while you can assign/edit/remove labels from a resource without affecting its liveness. These details become relevant since changing the labels on a running Akka node would make data inconsistent.

Regarding the deployment needs, I am 100% in favor of supporting monotonic revisions to identify the deployed version as metadata that's uncorrelated from the binary version (2.3.9-RC2.jar). I think a the `app-version` support Renato and Patrik mentioned is good enough for Kubernetes rolling update`s provided by `Deployment` but fall short in non-vanilla update strategies (canary/1-10-100/a-B) or even rollbacks.

**Comment by jroper** [comment]
@patriknw Yes I had heard about the new allocation strategy, I haven't had a chance to look at it closely but it sounds good.

Yes, we are planning to use a custom allocation strategy, and that's indeed what requesting this feature is all about.

@ignasi35 Annotations are also used by Kubernetes itself, and are used extensively by things like Istio, Dapr, and now Cloudstate to define sidecar injection semantics and configuration. cert-manager makes extensive use of annotations for certificate injection, and every ingress controller I've ever looked at heavily uses annotations. They are certainly not an application internal thing. The thing that differentiates labels from annotations is that labels are indexed in etcd (or at least, they will be at some point, they are designed with that intention) - which means you can efficiently look things up by a label. You can't look things up by an annotation, if you want to find every pod with a given annotation, you have to load every pod and check its annotations. Labels correspondingly have much tighter limits as to allowed characters, maximum lengths etc. So, where labels are used and where annotations are used depends on whether you need to look things up by the label - members of a service need to be looked up, so that uses labels, members of a replicaset need to be looked up, so that uses labels.

I'm asking for a generic feature because I have a vague idea of some other places that this might be useful - nothing that concrete, but I could imagine a number of different generic reusable cluster features/libraries that wish to attach some metadata to nodes.

**Comment by patriknw** [comment]
I think the need for this request is too vague and speculative so far.
I can imagine that if we add a full key-value Map on the Member that will be abused in ways that is not great for the membership gossip. Increasing message sizes. Distributed Data could be better for such Map.
